File: 0024. Swap Nodes in Pairs/main0024.java
import java.util.*;

public class main0024 {
    public static void main(String[] args) {
        Solution0024 sol = new Solution0024();
        int[] head;
        ListNode h1;

        head = new int[]{1,2,3,4};
        h1 = sol.intArr2List(head);
        sol.printList(h1);
        h1 = sol.swapPairsI(h1);        
        sol.printList(h1);
    }
}

class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

class Solution0024 {
    public ListNode intArr2List(int[] arr) {
        ListNode dum = new ListNode(0);
        ListNode cur = dum;
        for (int x : arr) {
            cur.next = new ListNode(x);
            cur = cur.next;
        }
        return dum.next;
    }
    public void printList(ListNode head) {
        List<Integer> res = new ArrayList<>();
        while (head != null) {
            res.add(head.val);
            head = head.next;
        }
        System.out.println(Arrays.toString(res.toArray()));
    }
    public ListNode swapPairs(ListNode head) {

        // If the list has no node or has only one node left.
        if ((head == null) || (head.next == null)) {
            return head;
        }

        // Nodes to be swapped
        ListNode firstNode = head;
        ListNode secondNode = head.next;

        // Swapping
        firstNode.next  = swapPairs(secondNode.next);
        secondNode.next = firstNode;

        // Now the head is the second node
        return secondNode;
    }
    
    
    public ListNode swapPairsI(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode dum = new ListNode(0);
        dum.next = head;
        ListNode cur, prev;
        prev = dum;
        cur = head;
        while (cur != null && cur.next != null) {
            prev.next = cur.next;
            prev = cur;
            cur = cur.next.next;
            prev.next.next = prev;
        }
        prev.next = cur;
        return dum.next;
    }
    
    public ListNode swapPairsII(ListNode head) {
        ListNode dummyHead = new ListNode(0);
        dummyHead.next = head;
        ListNode temp = dummyHead;
        while (temp.next != null && temp.next.next != null) {
            ListNode node1 = temp.next;
            ListNode node2 = temp.next.next;
            temp.next = node2;
            node1.next = node2.next;
            node2.next = node1;
            temp = node1;
        }
        return dummyHead.next;
    }
}

File: 0024. Swap Nodes in Pairs/main0024.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution0024:
    def swapPairs(self, head: ListNode) -> ListNode:        
        """
        :type head: ListNode
        :rtype: ListNode
        """
        # Dummy node acts as the prevNode for the head node
        # of the list and hence stores pointer to the head node.
        if not head or not head.next:
            return head
        
        first_node = head
        second_node = head.next

        first_node.next = self.swapPairs(second_node.next)
        second_node.next = first_node

        return second_node

    def swapPairsIter(self, head: ListNode) -> ListNode:
        dum = ListNode(0)
        dum.next = head
        prev = dum

        while head and head.next:
            # Nodes to be swapped
            first_node = head
            second_node = head.next

            # Swapping
            prev.next = second_node
            first_node.next = second_node.next
            second_node.next = first_node

            # Reinitializing the head and prev_node for next swap
            prev = first_node
            head = first_node.next
        
        return dum.next


    def lst2link(self, lst):
        cur = dummy = ListNode(0)
        for e in lst:
            cur.next = ListNode(e)
            cur = cur.next
        return dummy.next

    def print(self, lst):
        # defining a blank res variable
        res = ""
        # initializing ptr to head
        ptr = lst
       # traversing and adding it to res
        while ptr:
            res += str(ptr.val) + ", "
            ptr = ptr.next
       # removing trailing commas
        res = res.strip(", ")
        # chen checking if
        # anything is present in res or not
        if len(res):
            return "[" + res + "]"
        else:
            return "[]"

sol = Solution0024()
a1 = [1,2,3,4]   
h1 = sol.lst2link(a1)
print(sol.print(h1))
print("Recursion:")
res = sol.swapPairs(h1)
print(sol.print(res)) 
print("Iteration:")
h1 = sol.lst2link(a1)
res = sol.swapPairsIter(h1)
print(sol.print(res)) 

